asynciterator.ts

```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
   }
 }
 
+/**
+ * A synchronous mapping function from one element to another.
+ * The iterator performing the mapping is passed as a second argument.
+ */
+export type MapFunction<S, D = S> = (item: S) => D | null;
+
+/**
+ An iterator that calls a synchronous mapping function
+ on every item from its source iterator.
+ @extends module:asynciterator.AsyncIterator
+*/
+export class MappingIterator<S, D = S> extends AsyncIterator<D> {
+  private readonly _destroySource: boolean;
+  protected _source: InternalSource<S>;
+
+  /**
+   * Applies the given mapping to the source iterator.
+   */
+  constructor(
+    _source: AsyncIterator<S>,
+    private _map: MapFunction<S, D>,
+    options: SourcedIteratorOptions = {}
+  ) {
+    super();
+    this._source = _validateSource(_source);
+    this._destroySource = options.destroySource !== false;
+
+    if (_source.done) {
+      this.close();
+    }
+    else {
+      this._source._destination = this;
+      this._source.on('end', destinationClose);
+      this._source.on('error', destinationEmitError);
+      this._source.on('readable', destinationSetReadable);
+      // If we are given a source that is already readable
+      // then we need to set the state of this iterable to readable
+      // also as there is no guarantee that the is no forthcoming
+      // readable event from the source
+      this.readable = this._source.readable;
```

Comment:
Makes sense; can also set it to `true` should this cause trouble (it shouldn't).

Reply:
Setting it to `true` is fine in terms of correctness. In terms of performance, I think it is best to not be readable if the source is not readable as this avoids firing unnecessary `readable` events and in turn prevents any downstream iterators trying to read from this and getting `null` returned.

I think this may also allow a more marked perf. improvement when https://github.com/RubenVerborgh/AsyncIterator/pull/45 is re-addressed as that PR gets rid of a lot of unnecessary `read` calls and in many cases can just about halve the number of `.read()` called when taking unions over transformations.